### PR TITLE
[Breaking] Add multiclass and multi-label support to AUROC metric

### PR DESCRIPTION
**Pull request by manucouto1**
### Checklist

- [x] Confirmed that `cargo run-checks` command has been executed.
- [x] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Part of #544 (training metrics). Addresses the multiclass ROC AUC
extension raised in the issue thread.

### Changes

AUROC was restricted to binary classification and panicked on more
than two classes. This PR generalizes it:

- One-vs-Rest AUROC for binary, multiclass and multi-label, via the
  Mann–Whitney U statistic, aggregated with a `ClassReduction`
  (Macro = mean of per-class AUROC, Micro = pooled). `AurocMetric`
  holds a `ClassReduction` field directly (no extra config wrapper).
- `AurocMetric` now uses `ConfusionStatsInput`, so it is implicitly
  adapted for `ClassificationOutput` and `MultiLabelClassificationOutput`
  (same mechanism as Precision/Recall/FBetaScore).
- Public API aligned with the other classification metrics:
  `binary()`, `multiclass(class_reduction)`, `multilabel(class_reduction)`.
- Per-class degenerate columns (only positives or only negatives in the
  batch) are excluded from the Macro mean; if every class is degenerate
  the batch AUROC is undefined and reported as `0.5` (chance level) with
  a warning — rather than `NaN` (which would poison the running mean,
  since `NumericMetricState` aggregates without NaN filtering) or `0.0`
  (the worst score, misleading for an undefined batch).
- Removed the now-dead `AurocInput` type and its `Adaptor` impl.
- Book updated: AUROC added to `MultiLabelClassificationOutput`'s
  adapted metrics and to the `ConfusionStatsInput` footnote.

Review feedback addressed: dropped the `RankingMetricConfig` wrapper in
favour of a plain `ClassReduction` field; renamed the internal method to
`compute_auc`.

**Breaking:** `AurocInput` removed (use the built-in adaptors or
`ConfusionStatsInput`); the old no-arg `AurocMetric::new()` removed
(use `binary()`); metric name is now `AUROC [Macro|Micro]`.

Follow-up (out of scope): the pairwise implementation is O(n²·c) memory;
a sort-based O(n log n) version could replace it.

### Testing

- Parametrized `#[rstest]` suite (binary / multiclass / multi-label ×
  Micro/Macro) with inputs and expected values produced by an
  independent reference equivalent to scikit-learn's `roc_auc_score`
  (distinct scores, so the statistic is unambiguous); plus edge cases:
  perfect separation, chance level, Macro dropping a degenerate class,
  all-degenerate → 0.5, and name-by-reduction.
- `cargo test -p burn-train --lib metric::` passes; `cargo fmt --check`
  and `cargo clippy -p burn-train` are clean.
- The macOS CI failure is a pre-existing, flaky `burn-vision` / cubecl
  autotune panic unrelated to this change (the same `main` commit passes
  on re-run; this PR only touches `burn-train` metrics). Full
  `cargo run-checks` can't complete locally because the
  `wasm32-unknown-unknown` target isn't installed; CI covers the full
  matrix.

### CODE REVIEW

File: crates/burn-train/src/metric/auroc.rs [hunk 1]
```diff
@@ -2,100 +2,136 @@ use core::f64;
...
+            .bool_not()
+            .argwhere()
+            .squeeze_dim::<1>(1);
 
-        let area_under_curve = self.binary_auroc(&probabilities, &input.targets);
+        if keep.dims()[0] == 0 {
+            log::warn!(
+                "AUROC is undefined: no class has both positive and negative samples in the batch."
+            );
+            return 0.0;
```
Comment (nathanielsimard) [inline crates/burn-train/src/metric/auroc.rs]: We could return NaN or 0.5 instead of zero here.

Comment (manucouto1) [reply crates/burn-train/src/metric/auroc.rs]: Good catch. Went with 0.5 instead of NaN so one degenerate batch doesn't turn the epoch's running metric into NaN (NumericMetricState aggregates without NaN filtering).

File: crates/burn-train/src/metric/classification.rs [hunk 2]
```diff
@@ -7,6 +7,21 @@ pub struct ClassificationMetricConfig {
...
+}
+
+impl From<ClassificationMetricConfig> for RankingMetricConfig {
+    fn from(config: ClassificationMetricConfig) -> Self {
+        Self {
+            class_reduction: config.class_reduction,
+        }
+    }
+}
+
```
Comment (laggui) [inline crates/burn-train/src/metric/classification.rs]: I don't think this config is necessary. The `Auroc` can simply have a `ClassReduction` field instead, no?

File: crates/burn-train/src/metric/auroc.rs [hunk 3]
```diff
@@ -2,100 +2,137 @@ use core::f64;
...
+        let correct_pairs = reduce(si.clone().greater(sj.clone()).float() * valid.clone());
+        let tied_pairs = reduce(si.equal(sj).float() * valid);
 
         (correct_pairs + 0.5 * tied_pairs) / num_pairs
     }
-}
 
-impl Metric for AurocMetric {
-    type Input = AurocInput;
+    fn aucroc(&self, predictions: &Tensor<2>, targets: &Tensor<2, Bool>) -> f64 {
```
Comment (laggui) [inline crates/burn-train/src/metric/auroc.rs]: Small nit: would rename to something more meaningful e.g. `compute_auc` as it is an `AurocMetric` method anyway.

**Review by laggui — COMMENTED** [review]
Thanks for contributing!

Just a few minor comments

File: crates/burn-train/src/metric/classification.rs (hunk 2 above)
Comment (manucouto1) [reply crates/burn-train/src/metric/classification.rs]: Yeah, I agree — that’s simpler that way.

File: crates/burn-train/src/metric/auroc.rs (hunk 3 above)
Comment (manucouto1) [reply crates/burn-train/src/metric/auroc.rs]: Done — renamed to compute_auc.

**Review by laggui — APPROVED** [review]
